Panic during "zar find" on named field index

While the fix in #2273 appeared effective for the example in the original issue, it seems to have introduced a new problem, as this now causes a panic. Using repro steps similar to the `zar` README and testing with `zar` commit `77899a1` that's associated with the change in #2273:

```
$ zar -version
Version: v0.29.0-98-g77899a1f

$ rm -rf logs && mkdir ./logs && ZAR_ROOT=`pwd`/logs && zar import -s 25MB zng/*.gz && zar index create uri && zar find -f table uri=/file
added: file:///Users/phil/logs/zd/20180324/d-1pJ4Vycnq7cg5f5ZZBbr0JJZX6V.zng.zar/idx-1pJ4W3jpHPzoq35O0RxSd9Vg0ou.zng
added: file:///Users/phil/logs/zd/20180324/d-1pJ4VUdChU7I4ziQGZb9pgbARb4.zng.zar/idx-1pJ4W3jpHPzoq35O0RxSd9Vg0ou.zng
added: file:///Users/phil/logs/zd/20180324/d-1pJ4VfjUDKxjLlQdi7VTbW4pNyB.zng.zar/idx-1pJ4W3jpHPzoq35O0RxSd9Vg0ou.zng
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x168908c]

goroutine 50 [running]:
github.com/brimsec/zq/zng.(*Record).ZvalIter(...)
	/Users/phil/scratch/zq/zng/recordval.go:117
github.com/brimsec/zq/zng.(*Record).FieldIter(...)
	/Users/phil/scratch/zq/zng/recordval.go:124
github.com/brimsec/zq/expr.NewKeyCompareFn(0x0, 0xc0005581a0, 0x1, 0x1)
	/Users/phil/scratch/zq/expr/sort.go:100 +0x4c
github.com/brimsec/zq/microindex.NewFinderReader(0x1e20980, 0xc00055a680, 0xc00054a090, 0xc0005700a0, 0x4, 0x0, 0x0, 0x0, 0xc0005700a7, 0x0, ...)
	/Users/phil/scratch/zq/microindex/findreader.go:31 +0xf4
github.com/brimsec/zq/ppl/lake/index.FindFromPath(0x1e20980, 0xc00055a680, 0xc00054a090, 0xc0005700a0, 0x4, 0x0, 0x0, 0x0, 0xc0005700a7, 0x0, ...)
	/Users/phil/scratch/zq/ppl/lake/index/index.go:113 +0xad
github.com/brimsec/zq/ppl/lake/index.Find(0x1e20980, 0xc00055a680, 0xc00054a090, 0xc0005700a0, 0x4, 0x0, 0x0, 0x0, 0xc0005700a7, 0x0, ...)
	/Users/phil/scratch/zq/ppl/lake/index/index.go:109 +0x128
github.com/brimsec/zq/ppl/lake.Find.func2(0xc0005700a0, 0x4, 0x0, 0x0, 0x0, 0xc0005700a7, 0x0, 0xc0000e0140, 0x1c, 0x0, ...)
	/Users/phil/scratch/zq/ppl/lake/finder.go:105 +0x230
github.com/brimsec/zq/ppl/lake.Walk.func1(0x151eb225e9920000, 0x4e94914f0000, 0xc0002ee380, 0x4, 0x4, 0xc0002ee380, 0x4)
	/Users/phil/scratch/zq/ppl/lake/walk.go:160 +0x269
github.com/brimsec/zq/ppl/lake.tsDirVisit(0x1e20980, 0xc00055a680, 0xc00057e000, 0x0, 0x7fffffffffffffff, 0xc00011bce0, 0x0, 0x0)
	/Users/phil/scratch/zq/ppl/lake/walk.go:80 +0x3e2
github.com/brimsec/zq/ppl/lake.Walk(0x1e20980, 0xc00055a680, 0xc00057e000, 0xc0003e1e10, 0x0, 0x7ffeefbffa2a)
	/Users/phil/scratch/zq/ppl/lake/walk.go:154 +0x94
github.com/brimsec/zq/ppl/lake.Find(0x1e20980, 0xc00055a680, 0xc00054a090, 0xc00057e000, 0x0, 0x0, 0x7ffeefbffa2a, 0x3, 0x0, 0x0, ...)
	/Users/phil/scratch/zq/ppl/lake/finder.go:103 +0x412
github.com/brimsec/zq/ppl/cmd/zar/find.(*Command).Run.func1(0x1e20980, 0xc00055a680, 0xc00057e000, 0x0, 0x0, 0x7ffeefbffa2a, 0x3, 0x0, 0x0, 0xc0005581a0, ...)
	/Users/phil/scratch/zq/ppl/cmd/zar/find/command.go:123 +0x1d9
created by github.com/brimsec/zq/ppl/cmd/zar/find.(*Command).Run
	/Users/phil/scratch/zq/ppl/cmd/zar/find/command.go:122 +0x518
```

At commit 8fcbfbc that came right before that change, there was no panic. But maybe the incomplete results were just masking the problem?

```
$ zar -version
Version: v0.29.0-97-g8fcbfbc2

$ rm -rf logs && mkdir ./logs && ZAR_ROOT=`pwd`/logs && zar import -s 25MB zng/*.gz && zar index create uri && zar find -f table uri=/file
added: file:///Users/phil/logs/zd/20180324/d-1pJ4cF8luDX684PHIzfK2IfaUSd.zng.zar/idx-1pJ4cIfuv57n9svOzUditljUDHw.zng
added: file:///Users/phil/logs/zd/20180324/d-1pJ4bcpLmoGnUd5Gvm48mMsLa7s.zng.zar/idx-1pJ4cIfuv57n9svOzUditljUDHw.zng
added: file:///Users/phil/logs/zd/20180324/d-1pJ4brywUFcyPbzdZi1RlCNDANc.zng.zar/idx-1pJ4cIfuv57n9svOzUditljUDHw.zng
added: file:///Users/phil/logs/zd/20180324/d-1pJ4cCekSp3OrR6YJZ2a81WT4mq.zng.zar/idx-1pJ4cIfuv57n9svOzUditljUDHw.zng
KEY   COUNT _LOG                                                           FIRST                       LAST
/file 1     /Users/phil/logs/zd/20180324/d-1pJ4bcpLmoGnUd5Gvm48mMsLa7s.zng 2018-03-24T17:36:30.158766Z 2018-03-24T17:15:20.609736Z
/file 3     /Users/phil/logs/zd/20180324/d-1pJ4brywUFcyPbzdZi1RlCNDANc.zng 2018-03-24T17:36:30.01359Z  2018-03-24T17:15:20.600725Z
```

--- a/microindex/findreader.go
+++ b/microindex/findreader.go
@@ -15,6 +15,7 @@ import (
 type FinderReader struct {
 	compare expr.KeyCompareFn
 	finder  *Finder
+	inputs  []string
 	reader  zbuf.Reader
 }
 
@@ -23,27 +24,30 @@ func NewFinderReader(ctx context.Context, zctx *resolver.Context, uri iosrc.URI,
 	if err != nil {
 		return nil, err
 	}
-	keys, err := finder.ParseKeys(inputs...)
+	return &FinderReader{finder: finder, inputs: inputs}, nil
+}
+
+func (f *FinderReader) init() error {
+	keys, err := f.finder.ParseKeys(f.inputs...)
 	if err != nil {
-		finder.Close()
-		return nil, err
+		return err
 	}
-	compare, err := expr.NewKeyCompareFn(keys)
+	f.compare, err = expr.NewKeyCompareFn(keys)
 	if err != nil {
-		finder.Close()
-		return nil, err
+		return err
 	}
-	reader, err := finder.search(compare)
-	if err != nil {
-		finder.Close()
-		return nil, err
-	}
-	return &FinderReader{compare: compare, finder: finder, reader: reader}, nil
+	f.reader, err = f.finder.search(f.compare)
+	return err
 }
 
 func (f *FinderReader) Read() (*zng.Record, error) {
 	if f.finder.IsEmpty() {
 		return nil, nil
+	}
+	if f.compare == nil {
+		if err := f.init(); err != nil {
+			return nil, err
+		}
 	}
 	return lookup(f.reader, f.compare, f.finder.trailer.Order, eql)
 }

--- a/ztests/fixedbugs/2241.yaml
+++ b/ztests/fixedbugs/2241.yaml
@@ -1,0 +1,15 @@
+script: |
+  zar import -R ./zartest -
+  zar index create -R ./zartest uri
+  zar find -R ./zartest -f table uri=/file
+
+inputs:
+  - name: stdin
+    data: |
+      {_path:"conn",ts:2018-03-24T17:15:21.255387Z,uid:"C8Tful1TvM3Zf5x8fl" (bstring),id:{orig_h:10.164.94.120,orig_p:39681 (port=(uint16)),resp_h:10.47.3.155,resp_p:3389 (port)} (=0),proto:"tcp" (=zenum),service:null (bstring),duration:4.266ms,orig_bytes:97 (uint64),resp_bytes:19 (uint64),conn_state:"RSTR" (bstring),local_orig:null (bool),local_resp:null (bool),missed_bytes:0 (uint64),history:"ShADTdtr" (bstring),orig_pkts:10 (uint64),orig_ip_bytes:730 (uint64),resp_pkts:6 (uint64),resp_ip_bytes:342 (uint64),tunnel_parents:null (1=(|[bstring]|))}
+
+outputs:
+  - name: stdout
+    data: ""
+  - name: stderr
+    data: ""

